include/openthread/dns.h

```diff
@@ -140,6 +141,93 @@ void otDnsInitTxtEntryIterator(otDnsTxtEntryIterator *aIterator, const uint8_t *
  */
 otError otDnsGetNextTxtEntry(otDnsTxtEntryIterator *aIterator, otDnsTxtEntry *aEntry);
 
+/**
+ * This function is called when a DNS-SD query subscribes a service or service instance.
+ *
+ * @param[in] aContext      A pointer to the application-specific context.
+ * @param[in] aFullName     The null-terminated full service name (e.g. "_ipps._tcp.default.service.arpa."), or
+ *                          full service instance name (e.g. "OpenThread._ipps._tcp.default.service.arpa.").
+ *
+ */
+typedef void (*otDnssdQuerySubscribe)(void *aContext, const char *aFullName);
+
+/**
+ * This function is called when a DNS-SD query unsubscribes a service or service instance.
+ *
+ * @param[in] aContext      A pointer to the application-specific context.
+ * @param[in] aFullName     The null-terminated full service name (e.g. "_ipps._tcp.default.service.arpa."), or
+ *                          full service instance name (e.g. "OpenThread._ipps._tcp.default.service.arpa.").
+ *
+ */
+typedef void (*otDnssdQueryUnsubscribe)(void *aContext, const char *aFullName);
+
+/**
+ * This structure represents information of a discovered service instance for a DNS-SD query.
+ *
+ */
+typedef struct otDnssdServiceInstanceInfo
+{
+    const char *   mFullName;  ///< Full instance name (e.g. "OpenThread._ipps._tcp.default.service.arpa.").
+    const char *   mHostName;  ///< Host name.
+    otIp6Address   mAddress;   ///< Host IPv6 address.
```

Comment:
If there are more than one address associated with host, do we give the first one? if so may be good to mention it here?

Reply:
Should be better to allow multiple IP6 addresses. Will enhance.